### PR TITLE
Fix for flickering characters on NVIDIA GPUS for Tenchu Z

### PR DESCRIPTION
**Pull request by maksloboda**
This attempts to fix issues stated in https://github.com/xenia-project/game-compatibility/issues/592 for NVIDIA GPUS.
In this pr i add a custom parameter called `fuzzy_alpha_epsilon`. When it is set to zero everything works as it was before this pr. If it is set to a non zero value then it is used as a margin of error when checking alpha values.

Tenchu Z was tested with a value of 0.01.

This can possibly fix Perfect Dark and Culdcept SAGA but i cannot test them sadly.

### CODE REVIEW

File: src/xenia/gpu/dxbc_shader_translator_om.cc [hunk 1]
```diff
@@ -12,11 +12,17 @@
 #include <cstdint>
 
 #include "xenia/base/assert.h"
+#include "xenia/base/cvar.h"
 #include "xenia/base/math.h"
 #include "xenia/gpu/draw_util.h"
 #include "xenia/gpu/render_target_cache.h"
 #include "xenia/gpu/texture_cache.h"
 
+DEFINE_double(fuzzy_alpha_epsilon,
```
Comment (Gliniak) [inline src/xenia/gpu/dxbc_shader_translator_om.cc]: I strongly suggest to just define some constant epsilon in code and create boolean config value. 
This will prevent users from providing incorrect value and I expect every Nvidia GPU have exactly the same diff.

**Comment by Etokapa** [comment]
Tested on latest canary and this PR with a value of 0.01.

Improves Tenchu Z.
</img> </img> 

Improves Culdcept SAGA.
</img> </img> 

Improves Ninja Blade.
</img> </img> 

Improves Omega Five.
</img> </img> 

No noticable change for Perfect Dark.
</img> </img> 

Seems to remove the color from the boss, and noise is still present on REZ HD.
</img> </img> 

Overall, great work! :D

**Comment by Sowa95** [comment]
0.01 value, no change in UEFA Champions League 2006–2007.
 
 ucl

File: src/xenia/gpu/dxbc_shader_translator_om.cc (hunk 1 above)
Comment (maksloboda) [reply src/xenia/gpu/dxbc_shader_translator_om.cc]: Changed to bool in the new version and changed the name slightly, default epsilon is 0.01.

**Comment by Gliniak** [comment]
One more request. Please squash these commits into one and add prefix [GPU]

**Comment by TDOdev91** [comment]
Heya want to jump in and add that I found a fix for the shader bug on Perfect Dark XBLA for Nvidia GPU users but not sure how to implement it directly in a patch (never done any of this before lol).

I did a lot of research on the default.xex itself and after some time discovered that it was the `float4 FogNearFar: register(c1)` causing this shader noise to occur on all the models. This can be fixed simply by changing it to `float4 FogNearFar: register(c0)` and after that the game no longer has the issue with the shader bug. There have been some other bugs with this that I noted (I noticed the team colors in combat simulator only appear on characters after death but not on their person), but so far from what I have found they all happen to be quite minor. 

I had also tested this function to c(2) and this again fixes the shader bug, but best results seem to be at mitigating most bugs at the c(0) setting.

c1 setting
`float4 FogNearFar: register(c1)`

c0 setting
`float4 FogNearFar: register(c0)`

part 3
Specific address point in the default.xex where the function resides.

**Comment by Triang3l** [comment]
The proper fix for both Tenchu Z and Perfect Dark issues would be using manual vertex attribute interpolation using explicit barycentric coordinates as `v0 + i * (v1 - v0) + j * (v1 - v0)`, so that for equal per-vertex values, the barycentric coordinates will be canceled out (Perfect Dark has an exact `== -1.0` comparison for an interpolator input in the pixel shader, but due to how interpolation works on Nvidia, the value ends up slightly different from `-1.0`). However that's only possible on Turing GPUs and newer, and on Direct3D, would require a new shader translator that would generate DXIL instead of DXBC.

But this change is still conceptually applicable, as according to the specification of OpenGL, as well as the [specification of the `GL_QCOM_alpha_test` extension](https://registry.khronos.org/OpenGL/extensions/QCOM/QCOM_alpha_test.txt) used on the Adreno 200:

> `<ref>` is clamped to lie in [0,1], and then converted to a fixed-point value according to the rules given for an A component in section 2.12.8. For purposes of the alpha test, the fragment's alpha value is also rounded to the nearest integer.

So it appears that alpha test is done _after_ the conversion of the pixel shader color export from float32x4 to 32-bit or 64-bit, according to the color target 0 format, since it's apparently done by the render backend chip. Thus, for `8_8_8_8`, alpha test should be done with 8 bits of precision for both the alpha and the reference, rather than all 32. One interesting edge case though would be formats without alpha — `16_16` and `16_16_FLOAT` — but maybe for those the alpha should be assumed to be 1. The exact details need to be checked on the real Xenos, however.
